workflow/controller/operator.go

```diff
@@ -802,7 +802,7 @@ func (woc *wfOperationCtx) persistUpdates(ctx context.Context) {
 
 func (woc *wfOperationCtx) checkTaskResultsCompleted() bool {
 	taskResultsCompleted := woc.wf.Status.GetTaskResultsCompleted()
-	if len(taskResultsCompleted) == 0 {
+	if len(taskResultsCompleted) == 0 && woc.wf.Status.Outputs != nil {
```

Comment:
It does seem like we want to make sure there are actually Outputs.

But now it makes me wonder, when do the Outputs get set? Is it possible to arrive here when `Outputs==nil` because they haven't been set yet?

Reply:
@juliev0 Happy Holidays!

Outputs are first initialized and populated (at a node level) in `taskResultReconciliation()`. Outputs are subsequently initialized and populated to the global scope `wf.Status.Outputs` in `woc.podReconciliation()` via `woc.addOutputsToGlobalScope(newState.Outputs)`, assuming `Parameters` or `Artifacts` are defined for any node. Therefore, `Outputs==nil` when either `woc.TaskResultReconciliation()` hasn't been run yet, or `Parameters` or `Artifacts` aren't defined for any node in the workflow.

As long as `checkTaskResultsCompleted()` has been run after `taskResultReconciliation()`, `&& woc.wf.Status.Outputs != nil` is meaningful.

`checkTaskResultsCompleted()` is used in two places:

1) Before `woc.garbageCollectArtifacts(ctx)` (after `taskResultReconciliation()`)
```
	// Do artifact GC if all task results are completed.
	if woc.checkTaskResultsCompleted() {
		if err := woc.garbageCollectArtifacts(ctx); err != nil {
			woc.log.WithError(err).Error("failed to GC artifacts")
			return
		}
	} else {
		woc.log.Debug("Skipping artifact GC")
	}
```

2) Before `woc.deletetaskResults(ctx)` as part of `persistUpdates(ctx)` (called in conjunction with `woc.wf.Status.Phase.Completed()`, which ensures that the workflow has either succeeded, failed, or errored, and therefore after `checkTaskResultsCompleted()` has had a chance to run).
```
	// Make sure the TaskResults are incorporated into WorkflowStatus before we delete them.
	if woc.wf.Status.Phase.Completed() && woc.checkTaskResultsCompleted() {
		if err := woc.deleteTaskResults(ctx); err != nil {
			woc.log.WithError(err).Warn("failed to delete task-results")
		}
	}
```

Reply:
Happy holidays! 

Trying to follow. So, based on your statement about how "Outputs are subsequently initialized and populated to the global scope `wf.Status.Outputs` in `woc.podReconciliation()`", doesn't that mean that `Outputs` could also be nil in the case that `taskResultReconciliation()` had been run but `podReconciliation()` had not?

Reply:
Oops, I meant to say:

2) Before `woc.deletetaskResults(ctx)` as part of `persistUpdates(ctx)` (called in conjunction with `woc.wf.Status.Phase.Completed()`, which ensures that the workflow has either succeeded, failed, or errored, and therefore after **`woc.podReconciliation()`** has had a chance to run).

I missed the obvious (case 3), which I just introduced, inside `markWorkflowPhase`, which is called at the end of the operator loop under various conditions.
```
	switch phase {
	case wfv1.WorkflowSucceeded, wfv1.WorkflowFailed, wfv1.WorkflowError:
		// wait for all daemon nodes to get terminated before marking workflow completed & make sure task results are complete as well.
		if phase.Completed() && !woc.hasDaemonNodes() && woc.checkTaskResultsCompleted() {
			woc.log.Info("Marking workflow completed")
```

Good catch, yes, you are right. The case you identified can happen for `woc.garbageCollectArtifacts(ctx)` (case 1), which gets called right after `woc.taskResultReconciliation()`, but before `woc.podReconciliation()`. I think a race condition could occur where the `taskResultInformer` hasn't received an update yet, and therefore garbage collection is allowed to happen because wf.Status.Outputs == nil.

Since `woc.wf.Status.Outputs != nil` is meant to check that outputs are in global scope (for archiving purposes), maybe it's best that we pull that logic up to `if phase.Completed() && !woc.hasDaemonNodes() && woc.checkTaskResultsCompleted() {`.

For example:
```
	switch phase {
	case wfv1.WorkflowSucceeded, wfv1.WorkflowFailed, wfv1.WorkflowError:
		// wait for all daemon nodes to get terminated before marking workflow completed & make sure task results are complete as well.
		if phase.Completed() && !woc.hasDaemonNodes() && (woc.checkTaskResultsCompleted() || woc.wf.Status.Outputs == nil) {
			woc.log.Info("Marking workflow completed")
```

Reply:
The above doesn't work. I'm thinking through it and will propose another solution.

Reply:
@juliev0 I suggest the following changes. Notice the new `considerStatusOutputs bool` parameter for `checkTaskResultsCompleted`:

```
func (woc *wfOperationCtx) checkTaskResultsCompleted(considerStatusOutputs bool) bool {
	taskResultsCompleted := woc.wf.Status.GetTaskResultsCompleted()

	if considerStatusOutputs && woc.wf.Status.Outputs != nil && len(taskResultsCompleted) == 0 {
		return false
	}
	if !considerStatusOutputs && len(taskResultsCompleted) == 0 {
		return false
	}

	for _, completed := range taskResultsCompleted {
		if !completed {
			return false
		}
	}
	return true
}
```

Case 1)

```
	// Do artifact GC if all task results are completed.
	if woc.checkTaskResultsCompleted(false) {
		if err := woc.garbageCollectArtifacts(ctx); err != nil {
			woc.log.WithError(err).Error("failed to GC artifacts")
			return
		}
	} else {
		woc.log.Debug("Skipping artifact GC")
	}
```

Case 2)
```
	// Make sure the TaskResults are incorporated into WorkflowStatus before we delete them.
	if woc.wf.Status.Phase.Completed() && woc.checkTaskResultsCompleted(true) {
		if err := woc.deleteTaskResults(ctx); err != nil {
			woc.log.WithError(err).Warn("failed to delete task-results")
		}
	}
```

Case 3)
```
	switch phase {
	case wfv1.WorkflowSucceeded, wfv1.WorkflowFailed, wfv1.WorkflowError:
		// wait for all daemon nodes to get terminated before marking workflow completed & make sure task results are complete as well.
		if phase.Completed() && !woc.hasDaemonNodes() && woc.checkTaskResultsCompleted(true) {
			woc.log.Info("Marking workflow completed")
```

Reply:
Sorry, going back....let's start from "what was the original intention of this new check?"

Reply:
@juliev0 This is to ensure that the workflow is only marked as complete and archived once the ouputs have been reflected in the status.

Reply:
`len(taskResultsCompleted) == 0` can happen if there are no outputs for any nodes, so we can't always say that task results have not completed if `len(taskResultsCompleted) == 0`. If `wf.Status.Outputs` is defined and `len(taskResultsCompleted) == 0`, then we can say that task results have not completed.

If `wf.Status.Outputs` is defined (outputs or parameters do exist in at least one node), and `taskResultsComplete` is an empty map, then we know that the wait containers aren't done executing.

I should have been more clear with the above, the new usage of woc.CheckTaskResultsCompleted(true) in `case wfv1.WorkflowSucceeded, wfv1.WorkflowFailed, wfv1.WorkflowError:` is to prevent the workflow from archiving before the status is updated with the outputs.

Reply:
@juliev0 I just updated my last two comments since I did a really poor job in answering the first time.